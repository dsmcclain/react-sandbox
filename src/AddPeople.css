.AddPeople {
    margin: 20px;
    padding: 20px;
    border: 2px solid #ff999;
    background: #e46161;
    text-align: center;
}