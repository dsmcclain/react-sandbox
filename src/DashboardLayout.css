.DashboardLayout {
    display: flex;
    flex-direction: column;
    align-items: center;
    margin: 20px;
    padding: 20px;
    border: 2px solid #0F46FF;
    background: #80D0FF;
}
.DashboardLayout.container {
    align-self: stretch;
    flex-direction: column;
}
.DashboardLayout .editButton {
    align-self: flex-start;
}
.DashboardLayout .widget-rows {
    flex-direction: row;
    flex-wrap: wrap;
    align-self: stretch;
    display: inline-flex;
}