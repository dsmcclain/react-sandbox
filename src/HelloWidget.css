.HelloWidget {
    border: 2px solid #555;
    text-align: center;
    background: #FFFE90;
    color: #333;
    margin: 20px;
    padding: 20px;
  }