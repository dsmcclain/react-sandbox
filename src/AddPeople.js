import React, { Component } from 'react';
import './AddPeople.css';

class AddPeople extends Component {
    constructor(props) {
        super(props);
        this.state = { greetingName: '' };
        this.handleUpdate = this.handleUpdate.bind(this);
        this.addPeople = this.addPeople.bind(this);
    }

    addPeople() {
        this.props.addPeople(this.state.greetingName);
        this.setState({ greetingName: '' });
    }

    handleUpdate(e) {
        this.setState({ greetingName: e.target.value });
    }

    render() {
        return (
            <div className="AddPeople">
                <input 
                    type="text" 
                    onChange={this.handleUpdate}
                    value={this.state.greetingName}
                />
                &nbsp;&nbsp;
                <button onClick={this.addPeople}>Add</button>
            </div>
        );
    }
}

export default AddPeople;