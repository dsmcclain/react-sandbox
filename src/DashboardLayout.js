import React, { Component } from 'react';
import './DashboardLayout.css';
import AddPeople from './AddPeople';
import AddWidget from './AddWidget';

import WidgetFrame from './WidgetFrame';
import HelloWidget from './HelloWidget';
import TrackingWidget from './TrackingWidget';

class DashboardLayout extends Component {
    constructor(props) {
        super(props);
        this.state = { 
            people: ['Jim', 'Beth'],
            widgetTypes: ['Hello Widget', 'Fake Widget'],
            editable: false,
            search: "",
        };
        this.addPeople = this.addPeople.bind(this);
        this.onRemove = this.onRemove.bind(this);
        this.onEdit = this.onEdit.bind(this);
    }

    renderGreetings = (name) => {
        const {search} = this.state;

        if( search !== "" && name.toLowerCase().indexOf( search.toLowerCase() ) === -1 ){
            return null
        }

        return (
            <WidgetFrame 
                editable={this.state.editable} 
                title={name} 
                onRemove={this.onRemove}
                children = {
                    <HelloWidget 
                    key={name} 
                    name={name}
                    />
            } />
        );
    }

    renderTracking = () => {
        return (
            <WidgetFrame
                editable={this.state.editable}
                title="Tracked Terms"
                onRemove={this.onRemove}
                children = { <TrackingWidget/> }
            />
        );
    }

    onChange = e => {
        this.setState({ search: e.target.value });
    }

    addPeople(newName) {
        this.setState({ people: [...this.state.people, newName] });
    }

    onRemove(removeName) {
        const filteredPeople = this.state.people.filter(name => {
            return name !== removeName;
        });
        this.setState({ people:filteredPeople });
    }

    onEdit() {
        this.setState({
            editable: !this.state.editable,
        });
    }

    render() {
        return (
            <div className="DashboardLayout container">
                <div className="editButton">
                    <button  onClick={this.onEdit}>Edit</button>
                </div>
                <input placeholder="search..." onChange={this.onChange} />
                {this.state.editable && <AddWidget
                                         addPeople={this.addPeople} 
                                         />}
                <div className="widget-rows">
                    {
                        this.state.people.map(name => {
                            return this.renderGreetings(name)
                        })
                    }
                    {this.renderTracking()}
                </div>
            </div>
        );
    }
}

export default DashboardLayout;