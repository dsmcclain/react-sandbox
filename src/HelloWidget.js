import React, { Component } from 'react';
import './HelloWidget.css';

class HelloWidget extends Component {
  constructor(props) {
      super(props);
      this.state = { 
        greeting: 'Hello',
        AbbreviationCount: 0,
      };
      this.abbreviate = this.abbreviate.bind(this);
      this.removePeople = this.removePeople.bind(this);
  }

  abbreviate() {
      const fredness = this.state.greeting;
      if (fredness === 'Hello') {
        this.setState({ greeting: 'Hi' });
        this.setState({ AbbreviationCount: this.state.AbbreviationCount + 1 });
      } else {
        this.setState({ greeting: 'Hello' });
      };
  }

  removePeople() {
      this.props.removePeople(this.props.name);
  }
  
  render() {
    return (
      <div className="HelloWidget">
        {this.state.greeting} {this.props.name}!
        <br />
        This greeting has been abbreviated {this.state.AbbreviationCount} times.
        <br />
        <button className="abbreviate" onClick={this.abbreviate}>Abbreviate</button>
        <button className="remove" onClick={this.removePeople}>Remove Me!</button>
      </div>
    );
  }
}

export default HelloWidget;