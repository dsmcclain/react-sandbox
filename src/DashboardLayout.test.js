import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import DashboardLayout from './DashboardLayout';
import AddPeople from './AddPeople';
import HelloWidget from './HelloWidget';

describe(DashboardLayout, () => {
    const component = shallow(
        <DashboardLayout />
    );

    it('renders and matches our snapshot', () => {
        const component = renderer.create(
            <DashboardLayout />
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
    
    it('contains an AddPeople subcomponent', () => {
        expect(component.find(AddPeople)).toHaveLength(1);
    });

    it('contains the same number of HelloWidget components as people', () => {
        const helloWorlds = component.find(HelloWidget).length;
        const people   = component.state('people').length;
        expect(helloWorlds).toEqual(people);
    });

    it('adds another person when the addPeople function is called', () => {
        const before = component.find(HelloWidget).length;
        component.instance().addPeople('Sample');
        component.update();
        const after = component.find(HelloWidget).length;
        expect(after).toBeGreaterThan(before);
    });

    it('removes a person from the list when the removePerson function is called', () => {
        const before = component.find(HelloWidget).length;
        const removeMe = component.state('people')[0];
        component.instance().removePeople(removeMe);
        const after = component.find(HelloWidget).length;
        expect(after).toBeLessThan(before);
    });
})