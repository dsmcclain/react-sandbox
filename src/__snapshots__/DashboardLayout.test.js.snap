// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardLayout renders and matches our snapshot 1`] = `
<div
  className="DashboardLayout"
>
  <div
    className="AddPeople"
  >
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
      
    <button
      onClick={[Function]}
    >
      Add
    </button>
  </div>
  <div
    className="widget-frame"
  >
    <div
      className="widget-frame-header"
    >
      <span
        className="title"
      >
        Jim
      </span>
    </div>
    <div
      className="widget-frame-content"
    >
      <div
        className="HelloWidget"
      >
        Hello
         
        Jim
        !
        <br />
        This greeting has been abbreviated 
        0
         times.
        <br />
        <button
          className="abbreviate"
          onClick={[Function]}
        >
          Abbreviate
        </button>
        <button
          className="remove"
          onClick={[Function]}
        >
          Remove Me!
        </button>
      </div>
    </div>
  </div>
  <div
    className="widget-frame"
  >
    <div
      className="widget-frame-header"
    >
      <span
        className="title"
      >
        Beth
      </span>
    </div>
    <div
      className="widget-frame-content"
    >
      <div
        className="HelloWidget"
      >
        Hello
         
        Beth
        !
        <br />
        This greeting has been abbreviated 
        0
         times.
        <br />
        <button
          className="abbreviate"
          onClick={[Function]}
        >
          Abbreviate
        </button>
        <button
          className="remove"
          onClick={[Function]}
        >
          Remove Me!
        </button>
      </div>
    </div>
  </div>
</div>
`;
